src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java

```diff
@@ -0,0 +1,453 @@
+    public void apply(VariantContext variant, ReadsContext reads, ReferenceContext ref, FeatureContext features) {
+        final VariantContext newVC = regenotypeVC(ref, variant);
+        if (newVC != null) {
+            vcfWriter.add(newVC);
+        }
+    }
+
+    /**
+     * Re-genotype (and re-annotate) a VariantContext
+     *
+     * @param ref            the ref context
+     * @param originalVC     the combined genomic VC
+     * @return a new VariantContext or null if the site turned monomorphic and we don't want such sites
+     */
+    protected VariantContext regenotypeVC(final ReferenceContext ref, final VariantContext originalVC) {
+        if ( originalVC == null ) {
+            throw new IllegalArgumentException("originalVC cannot be null");
+        } else if (!isProperlyPolymorphic(originalVC) && !INCLUDE_NON_VARIANTS) {
+            return null;
+        }
+
+        VariantContext result = originalVC;
+
+        //Pass back ref-conf homRef sites/blocks to be combined by the GVCFWriter
+        if (isHomRefBlock(result)) {
+            return processHomRefBlock(result);
+        }
+
+        //don't need to calculate quals for sites with no data whatsoever or sites already genotyped homRef
+        if (result.getAttributeAsInt(VCFConstants.DEPTH_KEY,0) > 0 && !isHomRefCall(result)) {
+            final GenotypeLikelihoodsCalculationModel model = result.getType() == VariantContext.Type.INDEL
+                    ? GenotypeLikelihoodsCalculationModel.INDEL
+                    : GenotypeLikelihoodsCalculationModel.SNP;
+            result = genotypingEngine.calculateGenotypes(originalVC, model, null);
+        }
+
+        if (result == null || (!isProperlyPolymorphic(result) && !INCLUDE_NON_VARIANTS)) {
+            return null;
+        }
+
+        //variants with PL[0] less than threshold get turned to homRef with PL=[0,0,0], shouldn't get INFO attributes
+        //make sure we can call het variants with GQ >= RGQthreshold in joint calling downstream
+        if(shouldBeReblocked(result)) {
+            return reblockVariant(result, originalVC);
+        }
+        //high quality variant
+        else {
+            Map<String, Object> attrMap = new HashMap<>();
+            Map<String, Object> origMap = originalVC.getAttributes();
+            //pare down annotations to just VQSR features
+            for(final String annotation : annotationsToUse) {
+                if(origMap.containsKey(annotation))
+                    attrMap.put(annotation, origMap.get(annotation));
+            }
+            if (doQualApprox && result.getGenotype(0).hasPL()) {
+                attrMap.put(GATKVCFConstants.RAW_QUAL_APPROX_KEY, result.getGenotype(0).getPL()[0]);
+                int varDP = QualByDepth.getDepth(result.getGenotypes(), null);
+                if (varDP == 0) {  //prevent QD=Infinity case
+                    varDP = result.getAttributeAsInt(VCFConstants.DEPTH_KEY, Integer.MAX_VALUE);
```

Comment:
I guess that's fair.  My goal was to force QD to zero because people often hard filter on that and clearly this is a weird variant if it's missing all its depth. I really hate that we "correct" high QD values.  It's on my list.